integration/test/Test/Channels.hs

```diff
@@ -408,6 +411,111 @@ testTeamAdminCanCreateChannelWithoutJoining = do
+    I.getConversation channel `bindResponse` \resp -> do
+      resp.status `shouldMatchInt` 200
+      convMems <- resp.json %. "members" & asList
+      for [m1, m2, m3] (\m -> m %. "id") `shouldMatchSet` (for convMems (\m -> m %. "id"))
+
+    do
+      -- client if m1 receives the member join notification
+      notif <- awaitMatch isMemberJoinNotif ws1
+      -- if add_type is external_add ...
+      notif %. "payload.0.data.add_type" `shouldMatch` "external_add"
+      qconv <- notif %. "payload.0.qualified_conversation"
+      membersToAdd <- notif %. "payload.0.data.users" & asList
+
+      conv <- getConversation m1 qconv >>= getJSON 200
+      -- ... and the epoch is 0
+      conv %. "epoch" `shouldMatchInt` 0
+      -- the client creates the MLS group and adds everyone else
+      convId <- objConvId conv
+      createGroup def c1 convId
+      void $ createAddCommit c1 convId membersToAdd >>= sendAndConsumeCommitBundleExpectNoMemberJoin
+
+    -- the members that were added receive a welcome message
+    for_ [ws2, ws3] $ \ws -> do
+      awaitMatch isWelcomeNotif ws
+
+    -- the team admin adds another member to the channel
+    addMembers owner channel def {users = [m4, m5], role = Just "wire_member"} `bindResponse` \resp -> do
+      resp.status `shouldMatchInt` 200
+
+    notif <- awaitMatch isMemberJoinNotif ws4
+    notif %. "payload.0.data.add_type" `shouldMatch` "external_add"
+    qcid <- notif %. "payload.0.qualified_conversation"
+
+    -- c4 adds itself with an external add
+    conv <- getConversation m4 qcid >>= getJSON 200
+    convId <- objConvId conv
+    void $ createExternalCommit convId c4 Nothing >>= sendAndConsumeCommitBundleExpectNoMemberJoin
+    -- add others via normal commit
+    membersToAdd <- others m4 notif
+    void $ createAddCommit c4 convId membersToAdd >>= sendAndConsumeCommitBundleExpectNoMemberJoin
+    -- m5 receives welcome message
+    void $ awaitMatch isWelcomeNotif ws5
+  where
+    others self memberJoinNotif = do
+      allUsers <- memberJoinNotif %. "payload.0.data.users" & asList
+      selfQid <- self %. "qualified_id"
+      filterM (\m -> (/= selfQid) <$> (m %. "qualified_id")) allUsers
+
+sendAndConsumeCommitBundleExpectNoMemberJoin :: (HasCallStack) => MessagePackage -> App Value
+sendAndConsumeCommitBundleExpectNoMemberJoin = sendAndConsumeCommitBundleWithProtocolAndConsumer (const consumingMessagesExpectNoMemberJoin) MLSProtocolMLS
+
+consumingMessagesExpectNoMemberJoin :: (HasCallStack) => MessagePackage -> Codensity App ()
+consumingMessagesExpectNoMemberJoin mp = Codensity $ \k -> do
+  conv <- getMLSConv mp.convId
+  let oldClients = Set.delete mp.sender conv.members
+  let newClients = Set.delete mp.sender conv.newMembers
+  let clients =
+        map (,MLSNotificationMessageTag) (toList oldClients)
+          <> map (,MLSNotificationWelcomeTag) (toList newClients)
+
+  withWebSockets (map fst clients) $ \wss -> do
+    r <- k ()
```

Comment:
Hm, I'm wondering if it's okay to call the continuation before some other computations. :thinking: (I really don't know.)